tests/Feature/ConnectionsSetupTest.php

```diff
@@ -0,0 +1,287 @@
+
+            return;
+        }
+    }
+
+    public static function invalidConfigsProvider(): \Generator
+    {
+        yield 'url should be set always' => [
+            'defaultConnection' => 'libsql',
+            'config' => [
+                'driver' => 'libsql',
+                'url' => '', // empty url
+                'authToken' => '',
+                'syncUrl' => '',
+                'syncInterval' => 5,
+                'readYourWrites' => true,
+                'encryptionKey' => '',
+                'remoteOnly' => false,
+                'prefix' => '',
+                'database' => null,
+            ],
+            'error' => 'URL and database not set, please check your configuration',
+        ];
+        yield 'url should be correct for file' => [
+            'defaultConnection' => 'libsql',
+            'config' => [
+                'driver' => 'libsql',
+                'url' => 'file:test', // file config
+                'authToken' => '',
+                'syncUrl' => '',
+                'syncInterval' => 5,
+                'readYourWrites' => true,
+                'encryptionKey' => '',
+                'remoteOnly' => false,
+                'prefix' => '',
+                'database' => null,
+            ],
+            'error' => 'Got driver - file, please check your URL and driver config',
+        ];
+    }
+
+    public function testConnectionInMemory(): void
+    {
+        config()->set('database.default', 'libsql');
+        config()->set('database.connections.libsql', [
+            'driver' => 'libsql',
+            'url' => 'libsql::memory',
+            'authToken' => '',
+            'syncUrl' => '',
+            'syncInterval' => 5,
+            'readYourWrites' => true,
+            'encryptionKey' => '',
+            'remoteOnly' => false,
+            'prefix' => '',
+            'database' => null, // doesn't matter actually, since we use sqlite
+        ]);
+        // Get the default connection
+        $connection = DB::connection('libsql');
+
+        // Assert that the connection is an instance of LibSQLConnection
+        $this->assertInstanceOf(LibSQLConnection::class, $connection);
+
+        // Get the PDO instance
+        $pdo = $connection->getPdo();
+
+        $this->assertInstanceOf(LibSQLDatabase::class, $pdo);
+        $this->assertEquals('local', $pdo->getDb()->mode);
+        $this->assertEquals('memory', $pdo->getConnectionMode());
+        $result = $connection->select('PRAGMA database_list');
+        $this->assertNotEmpty($result);
+        $result = $pdo->query('SELECT sqlite_version()');
+        $this->assertNotEmpty($result, 'Failed to query SQLite version');
+    }
+
+    public function testConnectionLocalFile(): void
+    {
+        config()->set('database.default', 'sqlite');
+        config()->set('database.connections.local_file', [
+            'driver' => 'libsql',
+            'url' => 'libsql::file:tests/_files/test.db',
```

Comment:
What the reason of adding `libsql::file` prefix in `config.url` value? Did you really read the `libSQL` standard connection at [Turso Client PHP](https://github.com/tursodatabase/turso-client-php)?

Reply:
`What the reason of adding libsql::file prefix in config.url value?`

Reason is that ConfigurationUrlParser::parseConfiguration will parse this correctly into:
 - driver - libsql
 - database ($path) - file:tests/_files/test.db
 
 Then we look at the path:
 ```
elseif (str_starts_with($path, 'file:') !== false) {
    $this->connection_mode = 'local';
}
 ```
 And see that it's local connection.
 
 `Did you really read the libSQL standard connection at [Turso Client PHP](https://github.com/tursodatabase/turso-client-php)?`
 Guilty, not really. I've looked into examples, and then mostly looked into stubs, and your connection mode resolving, implemented and tested each of those (memory, local, remote, remote_replicated), but maybe it need more tests to support all DSN's.
I've tried to make URL's in laravel config to be close to  extension DSN one's.
https://github.com/tursodatabase/turso-driver-laravel/blob/main/src/libsql_php_extension.stubs.php


`Did you know if [Turso Client PHP](https://github.com/tursodatabase/turso-client-php) is not build under PDO?`
Nope, I don't know Rust, didn't look inside extension.

Reply:
Thank you to introduce `singleton` method, now the Service Provider look more clean and on point. But, still I need to tested with other Laravel instances and other Laravel Package. So far I worked on this driver to be able to work with the `stancl/tenancy` package.

```php
<?php

namespace Turso\Driver\Laravel;

use Illuminate\Database\DatabaseManager;
use Spatie\LaravelPackageTools\Package;
use Spatie\LaravelPackageTools\PackageServiceProvider;
use Turso\Driver\Laravel\Database\LibSQLConnection;
use Turso\Driver\Laravel\Database\LibSQLConnector;

class LibSQLDriverServiceProvider extends PackageServiceProvider
{
    public function configurePackage(Package $package): void
    {
        /*
         * This class is a Package Service Provider
         *
         * More info: https://github.com/spatie/laravel-package-tools
         */
        $package
            ->name('turso-driver-laravel');
    }

    public function register(): void
    {
        parent::register();

        $this->app->singleton(LibSQLConnector::class, function ($app) {
            return new LibSQLConnector();
        });

        $this->app->resolving('db', function (DatabaseManager $db) {
            $db->extend('libsql', function ($config, $name) {
                $config = config('database.connections.libsql');
                $config['name'] = $name;
                if (! isset($config['driver'])) {
                    $config['driver'] = 'libsql';
                }
                $db = app()->get(LibSQLConnector::class)->connect($config);

                $connection = new LibSQLConnection($db, $config['database'] ?? ':memory:', $config['prefix'], $config);
                app()->instance(LibSQLConnection::class, $connection);

                $connection->createReadPdo($config);

                return $connection;
            });
        });
    }
}

```

I love the idea about `getDb` and `getConnectionMode` function, but those method should be used in `LibSQLConnection.php`:

```php
public function getDb(): LibSQL
{
    return $this->db->getDb();
}

public function getConnectionMode(): string
{
    return $this->db->getConnectionMode();
}
```

It's should be available via `$this->db` property from the `__construct`:

```php
public function __construct(LibSQLDatabase $db, string $database = ':memory:', string $tablePrefix = '', array $config = [])
{
    
    $this->db = $db;
    $this->schemaGrammar = $this->getDefaultSchemaGrammar();
    $libsqlDatbase = function() use ($db) {
        return $db;
    };

    parent::__construct($libsqlDatbase, $database, $tablePrefix, $config);
}
```
Note: instead passing the `$db` variable into `Connection` constructor, now I used database `Closure`:

```diff
public function __construct(LibSQLDatabase $db, string $database = ':memory:', string $tablePrefix = '', array $config = [])
{
    
    $this->db = $db;
    $this->schemaGrammar = $this->getDefaultSchemaGrammar();
-   parent::__construct($db, $database, $tablePrefix, $config);
+   $libsqlDatbase = function() use ($db) {
+       return $db;
+   };
+
+   parent::__construct($libsqlDatbase, $database, $tablePrefix, $config);
}
```

`Illuminate\Database\Connection::construct` will Create a new database connection instance. And the first parameter should be `PDO` instance or `Closure`.

When those 2 methods `getDb` and `getConnecionMode` is used in `LibSQLConnection` then we can access it from `DB` Facades like `DB::getConnectionMode()`

Reply:
```
I love the idea about getDb and getConnectionMode function, but those method should be used in LibSQLConnection.php
```
It might be but, both LibSQLConnection and LibSQLDatabase has `db` property, which are actually different, so it will be very ambiguous to use it inside LibSQLConnection. 
I've added those methods to make classes more usable in tests only.